apps/nextjs/src/lib/lessonPlan/LessonPlanManager.ts

```diff
@@ -0,0 +1,70 @@
+import {
+  PatchDocument,
+  applyLessonPlanPatch,
+} from "@oakai/aila/src/protocol/jsonPatchProtocol";
+import { LooseLessonPlan } from "@oakai/aila/src/protocol/schema";
+import { Message } from "ai";
+import { createHash } from "crypto";
+import { deepClone } from "fast-json-patch";
+
+import { extractPatchesFromMessage } from "./extractPatches";
+
+export class LessonPlanManager {
+  private lessonPlan: LooseLessonPlan;
+  private appliedPatchHashes: Set<string> = new Set();
+
+  constructor(initialLessonPlan: LooseLessonPlan = {}) {
+    this.lessonPlan = deepClone(initialLessonPlan);
+  }
+
+  public getLessonPlan(): LooseLessonPlan {
+    return this.lessonPlan;
+  }
+
+  public applyPatchesFromMessage(message: Message): void {
+    if (message.role === "assistant") {
+      console.log("Applying patches from message", message.content);
+      const { validPatches } = extractPatchesFromMessage(message);
+      this.applyLocalPatches(validPatches);
+    }
+  }
+
+  public updateFromServer(serverLessonPlan: LooseLessonPlan): void {
+    // Update existing keys and add new ones
```

Comment:
This means we are maintaining a consistent reference to the lessonPlan which should mean fewer full re-renders

Reply:
I do think that this can make the rendering less predictable and I think we can solve the rerendering problem with smarter use of react techniques like context and state selectors

This mutation loop is already in our code so I think it's fine to keep it here for this PR at the moment